File: apps/native/Estates.Core/Secp256k1.cs
// Estates.Core/Secp256k1.cs — in-tree secp256k1 (NO third-party library). Pure C# over
// System.Numerics.BigInteger: field arithmetic mod p, affine point add/double/multiply,
// compressed point encode/decode, ECDH, and ECDSA (fresh CSPRNG random nonce k, low-S).
// This REPLACES BouncyCastle everywhere — the project depends on no crypto library.
// (SHA-256/512, HMAC, AES-GCM, HKDF are .NET framework primitives, not a library.)
using System.Numerics;
using System.Security.Cryptography;

namespace Estates.Core;

public static class Secp256k1
{
    // secp256k1 domain parameters (SEC 2).
    public static readonly BigInteger P = BigInteger.Parse("0FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F", System.Globalization.NumberStyles.HexNumber);
    public static readonly BigInteger N = BigInteger.Parse("0FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141", System.Globalization.NumberStyles.HexNumber);
    private static readonly BigInteger Gx = BigInteger.Parse("079BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798", System.Globalization.NumberStyles.HexNumber);
    private static readonly BigInteger Gy = BigInteger.Parse("0483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8", System.Globalization.NumberStyles.HexNumber);
    private static readonly BigInteger HalfN = N >> 1;

    public readonly record struct Point(BigInteger X, BigInteger Y, bool Inf)
    {
        public static readonly Point Identity = new(0, 0, true);
    }
    public static readonly Point G = new(Gx, Gy, false);

    private static BigInteger Mod(BigInteger a, BigInteger m) { var r = a % m; return r.Sign < 0 ? r + m : r; }
    private static BigInteger InvMod(BigInteger a, BigInteger m) => BigInteger.ModPow(Mod(a, m), m - 2, m); // p,n prime

    // ---- point arithmetic (affine) ----
    public static Point Add(Point p, Point q)
    {
        if (p.Inf) return q;
        if (q.Inf) return p;
        if (p.X == q.X)
        {
            if (Mod(p.Y + q.Y, P) == 0) return Point.Identity;   // p == -q
            return Double(p);
        }
        BigInteger s = Mod((q.Y - p.Y) * InvMod(q.X - p.X, P), P);
        BigInteger rx = Mod(s * s - p.X - q.X, P);
        BigInteger ry = Mod(s * (p.X - rx) - p.Y, P);
        return new Point(rx, ry, false);
    }

    public static Point Double(Point p)
    {
        if (p.Inf || p.Y == 0) return Point.Identity;
        BigInteger s = Mod((3 * p.X * p.X) * InvMod(2 * p.Y, P), P);   // a = 0
        BigInteger rx = Mod(s * s - 2 * p.X, P);
        BigInteger ry = Mod(s * (p.X - rx) - p.Y, P);
        return new Point(rx, ry, false);
    }

    // ---- FAST scalar multiply via JACOBIAN coordinates (X,Y,Z) ~ affine (X/Z², Y/Z³), a=0. This does ONE
    //      modular inverse at the very end instead of one per point-addition — ~10-50x faster than the affine
    //      double-and-add, which is what makes a 52-card multi-player mental-poker shuffle practical. Standard
    //      formulas (dbl-2009-l, madd-2007-bl); results are identical to the affine math (proven by the vectors).
    private static (BigInteger X, BigInteger Y, BigInteger Z) JDouble((BigInteger X, BigInteger Y, BigInteger Z) p)
    {
        if (p.Z.IsZero || p.Y.IsZero) return (1, 1, 0);   // infinity
        BigInteger A = Mod(p.X * p.X, P);
        BigInteger B = Mod(p.Y * p.Y, P);
        BigInteger C = Mod(B * B, P);
        BigInteger D = Mod(2 * (Mod((p.X + B) * (p.X + B), P) - A - C), P);
        BigInteger E = Mod(3 * A, P);
        BigInteger F = Mod(E * E, P);
        BigInteger X3 = Mod(F - 2 * D, P);
        BigInteger Y3 = Mod(E * (D - X3) - 8 * C, P);
        BigInteger Z3 = Mod(2 * p.Y * p.Z, P);
        return (X3, Y3, Z3);
    }
    private static (BigInteger X, BigInteger Y, BigInteger Z) JAddAffine((BigInteger X, BigInteger Y, BigInteger Z) p, Point q)
    {
        if (p.Z.IsZero) return (q.X, q.Y, 1);
        BigInteger Z1Z1 = Mod(p.Z * p.Z, P);
        BigInteger U2 = Mod(q.X * Z1Z1, P);
        BigInteger S2 = Mod(q.Y * p.Z * Z1Z1, P);
        BigInteger H = Mod(U2 - p.X, P);
        if (H.IsZero) { return Mod(S2 - p.Y, P).IsZero ? JDouble(p) : (1, 1, 0); }
        BigInteger HH = Mod(H * H, P);
        BigInteger I = Mod(4 * HH, P);
        BigInteger J = Mod(H * I, P);
        BigInteger r = Mod(2 * (S2 - p.Y), P);
        BigInteger V = Mod(p.X * I, P);
        BigInteger X3 = Mod(r * r - J - 2 * V, P);
        BigInteger Y3 = Mod(r * (V - X3) - 2 * p.Y * J, P);
        BigInteger Z3 = Mod(Mod((p.Z + H) * (p.Z + H), P) - Z1Z1 - HH, P);
        return (X3, Y3, Z3);
    }

    public static Point Mul(BigInteger k, Point p)
    {
        k = Mod(k, N);
        if (k.IsZero || p.Inf) return Point.Identity;
        var acc = ((BigInteger)0, (BigInteger)0, (BigInteger)0);   // infinity (Z=0)
        for (int i = (int)k.GetBitLength() - 1; i >= 0; i--)
        {
            acc = JDouble(acc);
            if (!((k >> i) & 1).IsZero) acc = JAddAffine(acc, p);
        }
        if (acc.Item3.IsZero) return Point.Identity;
        BigInteger zinv = InvMod(acc.Item3, P);
        BigInteger zinv2 = Mod(zinv * zinv, P);
        return new Point(Mod(acc.Item1 * zinv2, P), Mod(acc.Item2 * zinv2 * zinv, P), false);
    }

    // ---- scalars / encoding ----
    private static BigInteger Scalar(byte[] priv) => Mod(FromBytes(priv), N);
    private static BigInteger FromBytes(byte[] b) => new(Prepend0(b), isUnsigned: true, isBigEndian: true);
    private static byte[] Prepend0(byte[] b) { var o = new byte[b.Length + 1]; Array.Copy(b, 0, o, 1, b.Length); return o; }

    public static byte[] To32(BigInteger x)
    {
        byte[] b = x.ToByteArray(isUnsigned: true, isBigEndian: true);
        if (b.Length == 32) return b;
        var o = new byte[32];
        Array.Copy(b, 0, o, 32 - b.Length, Math.Min(b.Length, 32));
        return o;
    }

    /// <summary>Compressed (33-byte) public key for a 32-byte private key.</summary>
    public static byte[] PublicKey(byte[] priv) => Compress(Mul(Scalar(priv), G));

    public static byte[] Compress(Point p)
    {
        var o = new byte[33];
        o[0] = (byte)(p.Y.IsEven ? 0x02 : 0x03);
        Array.Copy(To32(p.X), 0, o, 1, 32);
        return o;
    }

    public static Point Decompress(byte[] pub)
    {
        if (pub.Length == 33 && (pub[0] == 0x02 || pub[0] == 0x03))
        {
            BigInteger x = FromBytes(pub[1..]);
            BigInteger y2 = Mod(BigInteger.ModPow(x, 3, P) + 7, P);
            BigInteger y = BigInteger.ModPow(y2, (P + 1) / 4, P);   // p ≡ 3 (mod 4)
            if (y.IsEven != (pub[0] == 0x02)) y = P - y;
            return new Point(x, y, false);
        }
        if (pub.Length == 65 && pub[0] == 0x04)
            return new Point(FromBytes(pub[1..33]), FromBytes(pub[33..]), false);
        throw new ArgumentException("bad public key encoding");
    }

    /// <summary>ECDH shared secret = compressed point (priv·peerPub).</summary>
    public static byte[] EcdhCompressed(byte[] priv, byte[] peerPub) => Compress(Mul(Scalar(priv), Decompress(peerPub)));
    /// <summary>ECDH shared secret = the X coordinate (32 bytes) of priv·peerPub.</summary>
    public static byte[] EcdhX(byte[] priv, byte[] peerPub) => To32(Mul(Scalar(priv), Decompress(peerPub)).X);

    // ---- MENTAL-POKER primitives (commutative encryption on the curve). A card i is the fixed public point
    //      Mᵢ = (i+1)·G; "encrypting" a card is scalar-multiplying its point; masks COMMUTE (a·(b·P)=b·(a·P)),
    //      so they strip in any order. These are thin wrappers over the field/point math above. ----

    /// <summary>The fixed PUBLIC base point for card index i: Mᵢ = (i+1)·G (compressed).</summary>
    public static byte[] CardBasePoint(int i)
    {
        if (i < 0) throw new ArgumentOutOfRangeException(nameof(i));
        return Compress(Mul(i + 1, G));
    }

    /// <summary>Scalar-multiply a compressed point by a 32-byte scalar: returns (scalar·P) compressed. This is
    /// one mask/unmask step. Throws on a malformed/off-curve point or a zero/invalid scalar — a hostile player
    /// cannot inject a bad point or a zero scalar into the deal.</summary>
    public static byte[] PointMul(byte[] point33, byte[] scalar)
    {
        if (!IsValidScalar(scalar)) throw new ArgumentException("invalid scalar");
        if (!IsValidPoint(point33)) throw new ArgumentException("invalid/off-curve point");
        return Compress(Mul(Scalar(scalar), Decompress(point33)));
    }

    /// <summary>The modular inverse (mod N) of a scalar, as 32 bytes — used to STRIP a mask (multiply by k⁻¹).
    /// Throws on a zero/invalid scalar (not invertible).</summary>
    public static byte[] ScalarInverse(byte[] scalar)
    {
        if (!IsValidScalar(scalar)) throw new ArgumentException("scalar not invertible (zero or out of range)");
        return To32(InvMod(FromBytes(scalar), N));
    }

    /// <summary>(a * b) mod N as 32 bytes — multiply two scalars (used to COMBINE per-card masks into one).
    /// Both inputs must be valid scalars in [1, N-1]; the product mod the prime N is never zero, so the result
    /// is always a valid, invertible scalar. Throws on a zero/out-of-range input.</summary>
    public static byte[] ScalarMul(byte[] a, byte[] b)
    {
        if (!IsValidScalar(a) || !IsValidScalar(b)) throw new ArgumentException("invalid scalar in ScalarMul");
        return To32(Mod(FromBytes(a) * FromBytes(b), N));
    }

    /// <summary>A scalar is valid iff it is 32 bytes encoding a value in [1, N-1].</summary>
    public static bool IsValidScalar(byte[] s)
    {
        if (s is not { Length: 32 }) return false;
        var x = FromBytes(s);
        return x > 0 && x < N;
    }

    /// <summary>A point is valid iff it is a 33-byte compressed point that lies on the curve and is not the
    /// point at infinity. Total — never throws.</summary>
    public static bool IsValidPoint(byte[] p)
    {
        try
        {
            if (p is not { Length: 33 } || (p[0] != 0x02 && p[0] != 0x03)) return false;
            var pt = Decompress(p);
            if (pt.Inf || pt.X <= 0 || pt.X >= P) return false;
            return Mod(pt.Y * pt.Y - (BigInteger.ModPow(pt.X, 3, P) + 7), P) == 0;   // y² = x³ + 7
        }
        catch { return false; }
    }

    // ---- ECDSA (fresh CSPRNG random nonce k, low-S). Each signature draws a new uniformly-random
    //      k in [1, n-1] from the OS CSPRNG via rejection sampling. ----
    private static BigInteger RandomK()
    {
        while (true)
        {
            BigInteger k = new(RandomNumberGenerator.GetBytes(32), isUnsigned: true, isBigEndian: true);
            if (k >= 1 && k < N) return k;
        }
    }

    /// <summary>Sign a 32-byte hash; returns (r, s) with low-S. The nonce k is freshly random.</summary>
    public static (BigInteger r, BigInteger s) SignHash(byte[] priv, byte[] hash32)
    {
        BigInteger d = Scalar(priv), z = Mod(FromBytes(hash32), N);
        while (true)
        {
            BigInteger k = RandomK();
            Point R = Mul(k, G);
            BigInteger r = Mod(R.X, N);
            if (r == 0) continue;
            BigInteger s = Mod(InvMod(k, N) * (z + r * d), N);
            if (s == 0) continue;
            if (s > HalfN) s = N - s;   // low-S
            return (r, s);
        }
    }

    /// <summary>Verify (r,s) over a 32-byte hash against a public key; rejects high-S.</summary>
    public static bool VerifyHash(byte[] pub, byte[] hash32, BigInteger r, BigInteger s)
    {
        try
        {
            if (r < 1 || r >= N || s < 1 || s >= N || s > HalfN) return false;
            BigInteger z = Mod(FromBytes(hash32), N);
            BigInteger w = InvMod(s, N);
            Point R = Add(Mul(Mod(z * w, N), G), Mul(Mod(r * w, N), Decompress(pub)));
            if (R.Inf) return false;
            return Mod(R.X, N) == r;
        }
        catch { return false; }
    }

    public static BigInteger ScalarAddModN(byte[] a, BigInteger b) => Mod(FromBytes(a) + b, N);
    public static BigInteger NOrder => N;
}

File: packages/keys/src/secp256k1.ts
/**
 * @estates/keys/secp256k1 — in-tree secp256k1, DEPENDENCY-FREE. The TypeScript twin of the
 * audited native `Secp256k1.cs`: same field arithmetic (mod p), affine point add/double/multiply,
 * compressed point encode/decode, ECDH, and ECDSA (CSPRNG random nonce k, low-S; NO RFC-6979).
 *
 * WHY: PLAN §1.2 — the product depends on NO third-party crypto library (`@noble/*` is banned, as
 * is `elliptic`). The runtime platform (WebCrypto / node:crypto) provides SHA-256/HMAC/HKDF/AES-GCM,
 * but NOT secp256k1, so the curve is implemented here, in-tree, identically to the native core.
 * Randomness is the platform CSPRNG (`crypto.getRandomValues`) — present in both browsers and Node.
 *
 * This file uses native `bigint` only; it imports nothing. It is the single source the rest of the
 * TS stack derives keys, signs protocol messages, and runs ECDH through (no Ed25519 anywhere).
 */

// ---- secp256k1 domain parameters (SEC 2) ----
export const P = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2Fn;
export const N = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141n;
const Gx = 0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798n;
const Gy = 0x483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8n;
const HALF_N = N >> 1n;

export interface Point { readonly x: bigint; readonly y: bigint; readonly inf: boolean }
const IDENTITY: Point = { x: 0n, y: 0n, inf: true };
export const G: Point = { x: Gx, y: Gy, inf: false };

function mod(a: bigint, m: bigint): bigint { const r = a % m; return r < 0n ? r + m : r; }
// modular inverse via Fermat (p, n are prime): a^(m-2) mod m
function powMod(base: bigint, exp: bigint, m: bigint): bigint {
  let r = 1n; base = mod(base, m);
  while (exp > 0n) { if (exp & 1n) r = (r * base) % m; base = (base * base) % m; exp >>= 1n; }
  return r;
}
function invMod(a: bigint, m: bigint): bigint { return powMod(mod(a, m), m - 2n, m); }

// ---- point arithmetic (affine) ----
export function add(p: Point, q: Point): Point {
  if (p.inf) return q;
  if (q.inf) return p;
  if (p.x === q.x) {
    if (mod(p.y + q.y, P) === 0n) return IDENTITY; // p == -q
    return double(p);
  }
  const s = mod((q.y - p.y) * invMod(q.x - p.x, P), P);
  const rx = mod(s * s - p.x - q.x, P);
  const ry = mod(s * (p.x - rx) - p.y, P);
  return { x: rx, y: ry, inf: false };
}

export function double(p: Point): Point {
  if (p.inf || p.y === 0n) return IDENTITY;
  const s = mod(3n * p.x * p.x * invMod(2n * p.y, P), P); // a = 0
  const rx = mod(s * s - 2n * p.x, P);
  const ry = mod(s * (p.x - rx) - p.y, P);
  return { x: rx, y: ry, inf: false };
}

// Scalar multiplication is the hot path. Doing it in AFFINE coordinates costs a modular
// inversion per add/double (~512 per multiply) — correct but far too slow. So `mul` runs in
// JACOBIAN projective coordinates (a:=0), where add/double are inversion-free, and converts to
// affine ONCE at the end (a single inversion per multiply). The affine result is identical, so
// `compress`/`decompress` and the KATs are unchanged. (a=0 ⇒ the standard "dbl-2009-l" / "add-2007-bl".)
interface Jac { X: bigint; Y: bigint; Z: bigint }
const JID: Jac = { X: 1n, Y: 1n, Z: 0n };

function jDouble(p: Jac): Jac {
  if (p.Z === 0n || p.Y === 0n) return JID;
  const A = mod(p.X * p.X, P);
  const B = mod(p.Y * p.Y, P);
  const C = mod(B * B, P);
  let D = mod((p.X + B) * (p.X + B) - A - C, P); D = mod(D + D, P);
  const E = mod(A + A + A, P);          // a = 0
  const F = mod(E * E, P);
  const X3 = mod(F - D - D, P);
  const Y3 = mod(E * (D - X3) - mod(C + C + C + C + C + C + C + C, P), P); // -8C
  const Z3 = mod((p.Y + p.Y) * p.Z, P);
  return { X: X3, Y: Y3, Z: Z3 };
}

function jAdd(p: Jac, q: Jac): Jac {
  if (p.Z === 0n) return q;
  if (q.Z === 0n) return p;
  const Z1Z1 = mod(p.Z * p.Z, P);
  const Z2Z2 = mod(q.Z * q.Z, P);
  const U1 = mod(p.X * Z2Z2, P);
  const U2 = mod(q.X * Z1Z1, P);
  const S1 = mod(p.Y * q.Z * Z2Z2, P);
  const S2 = mod(q.Y * p.Z * Z1Z1, P);
  if (U1 === U2) { return S1 === S2 ? jDouble(p) : JID; }
  const H = mod(U2 - U1, P);
  const I = mod((H + H) * (H + H), P);
  const J = mod(H * I, P);
  const r = mod((S2 - S1) * 2n, P);
  const V = mod(U1 * I, P);
  const X3 = mod(r * r - J - V - V, P);
  const Y3 = mod(r * (V - X3) - mod(S1 * J * 2n, P), P);
  const Z3 = mod((mod((p.Z + q.Z) * (p.Z + q.Z), P) - Z1Z1 - Z2Z2) * H, P);
  return { X: X3, Y: Y3, Z: Z3 };
}

function fromJac(j: Jac): Point {
  if (j.Z === 0n) return IDENTITY;
  const zinv = invMod(j.Z, P);
  const zinv2 = mod(zinv * zinv, P);
  return { x: mod(j.X * zinv2, P), y: mod(j.Y * zinv2 % P * zinv, P), inf: false };
}

export function mul(k: bigint, p: Point): Point {
  k = mod(k, N);
  if (k === 0n || p.inf) return IDENTITY;
  let r = JID;
  let addend: Jac = { X: p.x, Y: p.y, Z: 1n };
  while (k > 0n) {
    if (k & 1n) r = jAdd(r, addend);
    addend = jDouble(addend);
    k >>= 1n;
  }
  return fromJac(r);
}

// ---- scalars / encoding ----
const toHex = (b: Uint8Array): string => { let s = ''; for (const x of b) s += x.toString(16).padStart(2, '0'); return s; };
function fromHexStrict(h: string): Uint8Array {
  if (h.length % 2 !== 0 || /[^0-9a-fA-F]/.test(h)) throw new Error('secp256k1: invalid hex');
  const o = new Uint8Array(h.length / 2);
  for (let i = 0; i < o.length; i++) o[i] = parseInt(h.slice(i * 2, i * 2 + 2), 16);
  return o;
}
const bytesToBig = (b: Uint8Array): bigint => (b.length === 0 ? 0n : BigInt('0x' + toHex(b)));

// the canonical 32-byte big-endian encoder for a scalar/coordinate in [0, 2^256).
function be32(x: bigint): Uint8Array { return fromHexStrict((x & ((1n << 256n) - 1n)).toString(16).padStart(64, '0')); }
/** Canonical 32-byte big-endian encoding of a scalar/coordinate (public alias of be32). */
export function to32(x: bigint): Uint8Array { return be32(x); }

const scalar = (priv: Uint8Array): bigint => mod(bytesToBig(priv), N);

/** Compressed (33-byte) public key for a 32-byte private key. */
export function publicKey(priv: Uint8Array): Uint8Array { return compress(mul(scalar(priv), G)); }

export function compress(p: Point): Uint8Array {
  const o = new Uint8Array(33);
  o[0] = (p.y & 1n) === 0n ? 0x02 : 0x03;
  o.set(be32(p.x), 1);
  return o;
}

export function decompress(pub: Uint8Array): Point {
  if (pub.length === 33 && (pub[0] === 0x02 || pub[0] === 0x03)) {
    const x = bytesToBig(pub.slice(1));
    const y2 = mod(powMod(x, 3n, P) + 7n, P);
    let y = powMod(y2, (P + 1n) / 4n, P); // p ≡ 3 (mod 4)
    const wantEven = pub[0] === 0x02;
    if (((y & 1n) === 0n) !== wantEven) y = P - y;
    return { x, y, inf: false };
  }
  if (pub.length === 65 && pub[0] === 0x04) {
    return { x: bytesToBig(pub.slice(1, 33)), y: bytesToBig(pub.slice(33)), inf: false };
  }
  throw new Error('secp256k1: bad public key encoding');
}

/** ECDH shared secret = compressed point (priv·peerPub). */
export function ecdhCompressed(priv: Uint8Array, peerPub: Uint8Array): Uint8Array {
  return compress(mul(scalar(priv), decompress(peerPub)));
}
/** ECDH shared secret = the X coordinate (32 bytes) of priv·peerPub. */
export function ecdhX(priv: Uint8Array, peerPub: Uint8Array): Uint8Array {
  return be32(mul(scalar(priv), decompress(peerPub)).x);
}

/** True iff `priv` is a valid secp256k1 scalar (32 bytes, in [1, n-1]). */
export function isValidScalar(priv: Uint8Array | null | undefined): boolean {
  if (!priv || priv.length !== 32) return false;
  const d = bytesToBig(priv);
  return d > 0n && d < N;
}

/** True iff `pub` is a valid public key: decodable AND actually on the curve (y² ≡ x³+7 mod p),
 *  with both coordinates in the field. Rejects off-curve / malformed points fail-closed. */
export function isValidPublicKey(pub: Uint8Array): boolean {
  try {
    const pt = decompress(pub);
    if (pt.inf || pt.x <= 0n || pt.x >= P || pt.y <= 0n || pt.y >= P) return false;
    return mod(pt.y * pt.y - (pt.x * pt.x * pt.x + 7n), P) === 0n;
  } catch { return false; }
}

// ---- ECDSA (CSPRNG random nonce k, low-S). NO deterministic/RFC-6979 nonce. ----
function randomK(): bigint {
  // fresh, uniformly-random k in [1, n-1] from the platform CSPRNG, by rejection sampling.
  for (;;) {
    const b = new Uint8Array(32);
    crypto.getRandomValues(b);
    const k = bytesToBig(b);
    if (k >= 1n && k < N) return k;
  }
}

/** Sign a 32-byte hash; returns 64 bytes (r‖s) with low-S. The nonce k is freshly random. */
export function signHash(priv: Uint8Array, hash32: Uint8Array): Uint8Array {
  const d = scalar(priv), z = mod(bytesToBig(hash32), N);
  for (;;) {
    const k = randomK();
    const R = mul(k, G);
    const r = mod(R.x, N);
    if (r === 0n) continue;
    let s = mod(invMod(k, N) * (z + r * d), N);
    if (s === 0n) continue;
    if (s > HALF_N) s = N - s; // low-S
    const o = new Uint8Array(64);
    o.set(be32(r), 0); o.set(be32(s), 32);
    return o;
  }
}

/** Verify a 64-byte (r‖s) signature over a 32-byte hash; rejects high-S. Total: never throws. */
export function verifyHash(pub: Uint8Array, hash32: Uint8Array, sig: Uint8Array): boolean {
  try {
    if (sig.length !== 64) return false;
    const r = bytesToBig(sig.slice(0, 32)), s = bytesToBig(sig.slice(32));
    if (r < 1n || r >= N || s < 1n || s >= N || s > HALF_N) return false;
    const z = mod(bytesToBig(hash32), N);
    const w = invMod(s, N);
    const R = add(mul(mod(z * w, N), G), mul(mod(r * w, N), decompress(pub)));
    if (R.inf) return false;
    return mod(R.x, N) === r;
  } catch { return false; }
}

// ---- DER (strict, minimal) for BSV transaction signatures ----
function derInt(v: bigint): Uint8Array {
  let b = be32(v);
  let i = 0; while (i < b.length - 1 && b[i] === 0) i++;
  b = b.slice(i);
  if (b[0] & 0x80) { const o = new Uint8Array(b.length + 1); o.set(b, 1); return o; }
  return b;
}
/** Encode a 64-byte (r‖s) signature as a strict-DER signature (low-S preserved). */
export function derEncode(sig64: Uint8Array): Uint8Array {
  const r = derInt(bytesToBig(sig64.slice(0, 32))), s = derInt(bytesToBig(sig64.slice(32)));
  const body = [0x02, r.length, ...r, 0x02, s.length, ...s];
  return Uint8Array.from([0x30, body.length, ...body]);
}
/** Decode a strict-DER signature to (r, s) as a 64-byte buffer. Throws on malformed input. */
export function derDecode(der: Uint8Array): Uint8Array {
  let i = 0;
  if (der[i++] !== 0x30) throw new Error('der');
  i++; // total length
  if (der[i++] !== 0x02) throw new Error('der');
  const rl = der[i++]; const r = bytesToBig(der.slice(i, i + rl)); i += rl;
  if (der[i++] !== 0x02) throw new Error('der');
  const sl = der[i++]; const s = bytesToBig(der.slice(i, i + sl));
  const o = new Uint8Array(64); o.set(be32(r), 0); o.set(be32(s), 32);
  return o;
}

export const hexEncode = toHex;
export const hexDecode = fromHexStrict;
